firefox_ui_harness/runners/update.py

```diff
@@ -0,0 +1,76 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import shutil
+import tempfile
+
+import mozfile
+import mozinstall
+
+from base import FirefoxUITestRunner
+from ..testcases import UpdateTestCase
+
+
+DEFAULT_PREFS = {
+    'app.update.log': True,
+}
+
+
+class UpdateTestRunner(FirefoxUITestRunner):
+
+    def __init__(self, **kwargs):
+        FirefoxUITestRunner.__init__(self, **kwargs)
+
+        self.original_bin = self.bin
+
+        self.prefs.update(DEFAULT_PREFS)
+
+        # In case of overriding the update URL, set the appropriate preference
+        override_url = kwargs.pop('update_override_url', None)
+        if override_url:
+            self.prefs.update({'app.update.url.override': override_url})
+
+        self.run_direct_update = not kwargs.pop('update_fallback_only', False)
+        self.run_fallback_update = not kwargs.pop('update_direct_only', False)
+
+        self.test_handlers = [UpdateTestCase]
+
+    def duplicate_application(self, application_folder):
+        """Creates a copy of the specified binary."""
+        target_folder = tempfile.mkdtemp('.binary-update-tests')
+
+        self.logger.info('Creating a copy of the application at "%s".' % target_folder)
+        mozfile.remove(target_folder)
+        shutil.copytree(application_folder, target_folder)
+
+        return target_folder
+
+    def run_tests(self, tests):
+        source_folder = self.get_application_folder(self.original_bin)
+
+        def _run_tests(update_type):
+            target_folder = None
+
+            try:
+                target_folder = self.duplicate_application(source_folder)
+                self.bin = mozinstall.get_binary(target_folder, 'Firefox')
+
+                self.type = update_type
+                FirefoxUITestRunner.run_tests(self, tests)
+
+            except Exception:
```

Comment:
An exception could definitely happen in copying the application to the target location. Also I checked `run_tests` in Marionette and it could also raise exceptions on various places. Mainly when operating with the file system. There is no general raise/except block for that method. So I feel uncomfortable right now to trust this method.

Keep in mind that if direct updates are failing, we do not want to skip fallback updates but execute them however the direct updates ended-up.